File: 986.py
class Solution(object):
    def intervalIntersection(self, A, B):
        """
        :type A: List[List[int]]
        :type B: List[List[int]]
        :rtype: List[List[int]]
        """
        res=[]
        n,m=len(A),len(B)
        i,j=0,0
        while i<n and j<m:
            left=max(A[i][0],B[j][0])
            right=min(A[i][1],B[j][1])
            #判断是否有交集的条件
            if left<=right:
                res.append([left,right])
            #哪个右区间元素较小，指针就向前移动一位
            if A[i][1]<B[j][1]:
                i+=1
            else:
                j+=1   
        return res

File: 986.java
class Solution {
    public int[][] intervalIntersection(int[][] firstList, int[][] secondList) {

    ArrayList<int[]> list = new ArrayList<>();
    int i=0;
    int j=0;
    while (i<firstList.length&&j<secondList.length){
        int a1=firstList[i][0];
        int a2=firstList[i][1];
        int b1=secondList[j][0];
        int b2=secondList[j][1];
        if(b2>a2){
            i++;
        }else {
            j++;
        }
        if(b2>=a1&&a2>=b1){
            int left=Math.max(a1,b1);
            int right =Math.min(a2,b2);
            list.add(new int[]{left,right});
        }
    }
    return list.toArray(new int[list.size()][]);
    }
}